### PR TITLE
Add Typescript to `core-utils`

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
The big one...

This PR removes many `PropType`s throughout the packages. These should be replaced by their typescript alternatives

Removes all references to core-utils types. This make some of the propTypes in non typescripted packages a little more useless. This should be resolved by type-scripting them. 

Some stories are lost pending a new release of some of these packages. 

### Packages depending on updated `from-to-picker`
- park-and-ride-overlay
- stops-overlay
- vehicle-rental-overlay

I think my strategy here will be to split out the `core-utils` and `from-to-picker` parts of this PR into a new one, and then have that PR block this one.

### CODE REVIEW

**Comment by miles-grant-ibigroup** [comment]
I think something weird may have happened with the snapshots... Would appreciate a second eye on this

**Review by philip-cline — APPROVED** [review]
This seemed... not fun to pull off..  so thank you so much for getting this done. Looking good to me, just with maybe a few more comments where typescripting still needs to be completed.

File: packages/icons/src/leg-icon.js [hunk 1]
```diff
@@ -21,7 +21,8 @@ const LegIcon = ({ getCompanyIcon, leg, ModeIcon, ...props }) => {
 LegIcon.propTypes = {
   // Optional override function for deriving the company icon for a given leg.
   getCompanyIcon: PropTypes.func,
-  leg: coreUtils.types.legType.isRequired,
+  // TYPESCRIPT TODO: restore
+  // leg: coreUtils.types.legType.isRequired,
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/icons/src/leg-icon.js:25]: I guess since we're basically disabling the react PropTypes it doesn't matter but surely better to have the correct object subtype where possible! Since it doesn't come from core-utils I think it should be fine right

File: packages/route-viewer-overlay/src/index.js [hunk 2]
```diff
@@ -142,7 +141,7 @@ RouteViewerOverlay.propTypes = {
     color: PropTypes.string,
     patterns: PropTypes.objectOf(
       PropTypes.shape({
-        geometry: coreUtils.types.encodedPolylineType,
+        // geometry: coreUtils.types.encodedPolylineType,
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/route-viewer-overlay/src/index.js:144]: Good catch thank you! This might've been lost to the ages otherwise

File: packages/stop-viewer-overlay/src/StopViewerOverlay.story.js [hunk 3]
```diff
@@ -23,7 +26,8 @@ function CustomMarker({ stop }) {
 }
 
 CustomMarker.propTypes = {
-  stop: coreUtils.types.stopLayerStopType.isRequired
+  // stop: coreUtils.types.stopLayerStopType.isRequired
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/stop-viewer-overlay/src/StopViewerOverlay.story.js:29]: Good catch thank you! This might've been lost to the ages otherwise

File: packages/core-utils/src/route.ts [hunk 4]
```diff
@@ -151,9 +160,13 @@ function makeTransitOperatorComparator(transitOperators) {
  * Gets the desired sort values according to an optional getter function. If the
  * getter function is not defined, the original sort values are returned.
  */
-function getSortValues(getterFn, a, b) {
-  let aVal;
-  let bVal;
+function getSortValues(
+  getterFn: (item: unknown) => unknown,
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/route.ts:164]: Wow this needs TypeScript FIXME comments...

File: packages/core-utils/src/route.ts [hunk 5]
```diff
@@ -220,15 +234,15 @@ function getRouteTypeComparatorValue(route) {
...
+function routeTypeComparator(a: Route, b: Route): number {
   return getRouteTypeComparatorValue(a) - getRouteTypeComparatorValue(b);
 }
 
 /**
  * Determines whether a value is a string that starts with an alphabetic
  * ascii character.
  */
-function startsWithAlphabeticCharacter(val) {
+function startsWithAlphabeticCharacter(val: unknown): boolean {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/route.ts:245]: Add at least a FIXME. I think route is `shortName ?: string | number` but don't see an existing definition.

File: packages/core-utils/src/route.ts [hunk 6]
```diff
@@ -390,7 +412,9 @@ function makeMultiCriteriaSort(...criteria) {
  *    those with shortNames.
  *  7. longName as string.
  */
-export function makeRouteComparator(transitOperators) {
+export function makeRouteComparator(
+  transitOperators: TransitOperator[]
+): (a: unknown, b: unknown) => number {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/route.ts:417]: Aren't `a` and `b` of type `Route`?

File: packages/types/src/index.ts [hunk 7]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
+  timezone?: string;
+  lang?: string;
+  phone?: string;
+  fareUrl?: string;
+};
+export type Route = {
+  agency: Agency;
+  agencyId?: string;
+  agencyName?: string;
+  shortName: string;
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:450]: Should be `: string | number` or `?: string | number`. See uses in the route comparator.

File: packages/types/src/index.ts [hunk 8]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
+  longName: string;
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:451]: Should be optional.

File: packages/types/src/index.ts [hunk 9]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
+  mode?: string;
+  id?: string;
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:453]: Isn't id required?

File: packages/core-utils/src/time.ts [hunk 10]
```diff
@@ -131,7 +146,7 @@ export function offsetTime(ms, options) {
  * @param {number} ms epoch time value in milliseconds
  * @returns {string} formatted text representation
  */
-export function formatTime(ms, options) {
+export function formatTime(ms: number, options: TimeOptions): string {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/time.ts:149]: We should deprecate this method.

File: packages/core-utils/src/time.ts [hunk 11]
```diff
@@ -113,8 +121,15 @@ export function formatDuration(seconds) {
  * @param {object} region  an object that allows internationalization of the time
  * @returns {string}       formatted text representation
  */
-export function formatDurationWithSeconds(seconds, region) {
-  return formatDurationLikeMoment(seconds, { enabled: true, code: region });
+// TS TODO: region as type?
+export function formatDurationWithSeconds(
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/time.ts:125]: We should deprecate this method.

File: packages/core-utils/src/time.ts [hunk 12]
```diff
@@ -103,8 +107,12 @@ export function getLongDateFormat(config) {
  * @param {number} seconds duration in seconds
  * @returns {string} formatted text representation
  */
-export function formatDuration(seconds) {
-  return formatDurationLikeMoment(seconds, false);
+// TS TODO: region as type?
+export function formatDuration(seconds: number, region: string): string {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/time.ts:111]: We should deprecate this method (implementers now use formatjs)

File: packages/endpoints-overlay/package.json [hunk 13]
```diff
@@ -9,7 +9,7 @@
   "homepage": "https://github.com/opentripplanner/otp-ui#readme",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "tsc": "tsc && echo Remove the @ts-expect-error exception in index.tsx once core-utils is typed!"
```
Comment (binh-dam-ibigroup) [inline packages/endpoints-overlay/package.json:12]: I would wait for the new core-utils version number before updating this package.

File: packages/icons/src/leg-icon.js [hunk 14]
```diff
@@ -21,7 +21,8 @@ const LegIcon = ({ getCompanyIcon, leg, ModeIcon, ...props }) => {
 LegIcon.propTypes = {
```
Comment (binh-dam-ibigroup) [inline packages/icons/src/leg-icon.js:21]: Is the better approach to convert this package to TypeScript?

File: packages/itinerary-body/src/styled.tsx [hunk 15]
```diff
@@ -302,13 +302,17 @@ export const LegDescriptionHeadsignPrefix = styled.span`
  * Lets others apply styles to the mode text in
  * "Bicycle 0.5 miles to City Hall"
  */
-export const LegDescriptionMode = styled.span``;
+export const LegDescriptionMode = styled.span`
+  font-weight: inherit;
```
Comment (binh-dam-ibigroup) [inline packages/itinerary-body/src/styled.tsx:306]: Unrelated change, please revert and move to another PR (2 instances).

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
First round of comments...

File: packages/types/src/index.ts [hunk 16]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
 /**
- * @deprecated
+ * This models data about a stop and it's associated routes that is obtained
+ * from a transit index API.
+ */
+export type Stop = {
+  /**
+   * The stop code if the stop has one
+   */
+  code: string;
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:429]: Make attribute optional to not contradict the comment above.

File: packages/types/src/index.ts [hunk 17]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
+  dist: number;
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:430]: Is this optional too?

File: packages/types/src/index.ts [hunk 18]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
+  shortName: string;
+  longName: string;
+  mode?: string;
+  id?: string;
+  // TS TODO: route type enum
+  type?: number;
+  color?: string;
+  textColor?: string;
+  routeBikesAllowed?: number;
+  bikesAllowed?: number;
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:459]: Are these number enums? (e.g. `0 | 1`)

File: packages/types/src/index.ts [hunk 19]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
+  sortOrder: number;
+  eligibilityRestricted?: number;
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:461]: Is this a number enum? (e.g. `0 | 1`)

File: packages/types/src/index.ts [hunk 20]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
+  offset: number;
+};
+
+export type TransitivePlace = {
+  placeId: string;
+  type: string;
+};
+
+export type TransitiveData = {
+  journeys: {
```
Comment (binh-dam-ibigroup) [inline packages/types/src/index.ts:488]: Extract a type `TransitiveJourney`. Extract other nested types too.

**Review by binh-dam-ibigroup — COMMENTED** [review]
Last round of comments for now. Can you make release-triggering package changes to non core-utils packages to another PRs, please.

File: packages/base-map/src/index.js [hunk 21]
```diff
@@ -332,7 +331,10 @@ BaseMap.propTypes = {
   /**
    * The center of the map, as a [lat, lng] array.
    */
-  center: utils.types.latlngType.isRequired,
+  // center: utils.types.latlngType.isRequired,
+  // Typescript TODO: restore correct type
+  // eslint-disable-next-line react/forbid-prop-types
+  center: PropTypes.object.isRequired,
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.js:337]: My approach for solving this problem was commenting out things here as needed, NOT publishing a package release and then having the typescript release be the next release for `base-map` (and the other packages in the same situation)

File: packages/types/src/index.ts [hunk 22]
```diff
@@ -8,18 +8,18 @@ import React, { FunctionComponent, ReactElement } from "react";
 /**
  * Shape for a transportation company.
  */
-export interface Company {
+export type Company = {
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts:11]: I had better success with autocomplete inference when using `type` and couldn't see a reason to stick with `interface`

File: packages/types/src/index.ts [hunk 23]
```diff
@@ -372,37 +375,258 @@ export interface Itinerary {
...
-  name: string;
+  name?: string;
   /**
    * This is only used location that a user has saved. Can be either:
-   * "home" or "work"
+   * One of: 'home', 'work', 'stop' or 'recent'
    */
-  type?: string;
-}
+  type?: "home" | "work" | "stop" | "recent" | string;
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts:400]: Good call. Removed!

File: packages/core-utils/src/itinerary.ts [hunk 24]
```diff
@@ -216,7 +228,7 @@ export function getCompanyFromLeg(leg) {
   return null;
 }
 
-export function getItineraryBounds(itinerary) {
+export function getItineraryBounds(itinerary: Itinerary): number[] {
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.ts:231]: Good idea!

File: packages/core-utils/src/route.ts (hunk 4 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/route.ts:164]: I can add a comment but they were made `unknown` explicitly. This method does not know the contents of what it is receiving as it's designed to accept anything

File: packages/core-utils/src/route.ts (hunk 5 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/route.ts:245]: I think here again an explicit `unknown` is ok. If you were to pass say a function in, this method would correctly return `false`

File: packages/core-utils/src/route.ts (hunk 6 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/route.ts:417]: They are `number`s. Good catch thank you!

File: packages/itinerary-body/src/styled.tsx (hunk 15 above)
Comment (miles-grant-ibigroup) [reply packages/itinerary-body/src/styled.tsx:306]: This is needed for the snapshots to pass

File: packages/endpoints-overlay/package.json (hunk 13 above)
Comment (miles-grant-ibigroup) [reply packages/endpoints-overlay/package.json:12]: I agree. This is why I haven't made a fix or feat commit for these packages.

File: packages/icons/src/leg-icon.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/icons/src/leg-icon.js:25]: It is not ideal, which is why I'd prefer not to make a package release with it! Instead, I'd prefer to typescript these packages next and then make a release from that

File: packages/base-map/src/index.js [hunk 25]
```diff
@@ -379,7 +381,10 @@ BaseMap.propTypes = {
    */
   popup: PropTypes.shape({
     contents: PropTypes.node.isRequired,
-    location: utils.types.latlngType.isRequired
+    // Typescript TODO: restore correct type
+    // eslint-disable-next-line react/forbid-prop-types
+    location: PropTypes.object.isRequired
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/base-map/src/index.js:386]: `latLngType` added in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/types/src/index.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts:402]: Some parts of the code insert this where strings are needed, and typescript is unhappy unless this type is also a string

File: packages/core-utils/src/itinerary.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.ts]: Good call. Refactored in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/itinerary.ts [hunk 26]
```diff
@@ -264,12 +276,16 @@ export function legLocationAtDistance(leg, distance) {
...
+export function legElevationAtDistance(
+  points: number[],
+  distance: number
+): number[] {
   // Iterate through the combined elevation profile
   let traversed = 0;
   // If first point distance is not zero, insert starting point at zero with
   // null elevation. Encountering this value should trigger the warning below.
   if (points[0][0] > 0) {
+    // @ts-expect-error TODO: how does this work??
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.ts]: Yes that fixed it thank you! Had to change the return type as well. db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/itinerary.ts [hunk 27]
```diff
@@ -349,10 +368,12 @@ export function getElevationProfile(steps, unitConversion = 1) {
...
  * @see https://stackoverflow.com/questions/118241/calculate-text-width-with-javascript/21015393#21015393
  */
-export function getTextWidth(text, font = "22px Arial") {
+export function getTextWidth(text: string, font = "22px Arial"): number {
   // re-use canvas object for better performance
   const canvas =
+    // @ts-expect-error TODO: what is this doing? How does it work?
     getTextWidth.canvas ||
+    // @ts-expect-error TODO: what is this doing? How does it work?
     (getTextWidth.canvas = document.createElement("canvas"));
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/itinerary.ts]: That worked great thanks so much! db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/map.ts [hunk 28]
```diff
@@ -298,27 +315,30 @@ export function itineraryToTransitive(
...
   return Number.isFinite(lat) && lat >= -90 && lat <= 90;
 }
 
-export function isValidLng(lng) {
+export function isValidLng(lng: number): boolean {
   return Number.isFinite(lng) && lng >= -180 && lng <= 180;
 }
 
-export function isValidLatLng(arr) {
+export function isValidLatLng(arr: [number, number]): boolean {
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/map.ts]: Refactored in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/map.ts [hunk 29]
```diff
@@ -298,27 +315,30 @@ export function itineraryToTransitive(
...
   return (
     Array.isArray(arr) &&
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/map.ts:344]: I think it can't hurt -- we can't guarantee that the user is using typescript, so this check which avoids crashes feels safe

File: packages/types/src/index.ts (hunk 7 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts:457]: Good catch thanks! Refactored in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/types/src/index.ts (hunk 8 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts]: Good catch! Refactored in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/types/src/index.ts (hunk 9 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts]: Perhaps? Refactored in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/time.ts (hunk 10 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/time.ts]: Deprecated in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/time.ts (hunk 11 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/time.ts]: Deprecated in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/time.ts (hunk 12 above)
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/time.ts]: Deprecated in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/core-utils/src/route.ts [hunk 30]
```diff
@@ -345,8 +365,10 @@ function getRouteSortOrderValue(val) {
  * returned. If all comparison functions return equivalence, then the values
  * are assumed to be equivalent.
  */
-function makeMultiCriteriaSort(...criteria) {
-  return (a, b) => {
+function makeMultiCriteriaSort(
+  ...criteria: ((a: unknown, b: unknown) => number)[]
```
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/core-utils/src/route.ts:369]: It is not known what they are! Part of what the method does is to figure that out

File: packages/types/src/index.ts (hunk 16 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts]: Good catch! Refactored in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/types/src/index.ts (hunk 17 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts]: I think so. Refactored in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/types/src/index.ts (hunk 18 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts]: I like this! Added in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/types/src/index.ts (hunk 19 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts]: Yep replaced in db1658fad93479daa48f7229371b9aef2200ae86

File: packages/types/src/index.ts (hunk 20 above)
Comment (miles-grant-ibigroup) [reply packages/types/src/index.ts]: Much cleaner thanks! in db1658fad93479daa48f7229371b9aef2200ae86

**Comment by miles-grant-ibigroup** [comment]
There is still some discussion to be had on how to handle the package releases of the packages that depend on the core-utils update. Since the snapshots are generated using the latest package release rather than the version in a package's `package.json`, I can't think of a way to avoid the changes here without deleting snapshots

**Review by binh-dam-ibigroup — COMMENTED** [review]
https://github.com/opentripplanner/otp-ui/pull/367/files/4d08f69c81ec924f497e0d95427c0a505cffc674#r874820444 still needs to be handled.

**Comment by miles-grant-ibigroup** [comment]
The PropTypes error is resolved in 0fd1509
